### ISSUE TITLE
Fuzzy matcher not finding best match!

### ISSUE THREAD

**Issue by linas** [opencog/opencog#1916]
Playing with the chatbot, it seems like the fuzzy matcher can get the right answer, but then forgets it ... sample from IRC:
```

<linas> cogita-bot: The color of the book is red.
<cogita-bot> Hello linas, you said: " The color of the book is red."
<cogita-bot> You made a Declarative SpeechAct
<linas> cogita-bot: What is the color of the book?
<cogita-bot> Hello linas, you said: " What is the color of the book?"
<cogita-bot> You made an Interrogative SpeechAct
<cogita-bot> the color of the book is red .
<linas> OK, that was slow, but it got it
<linas> cogita-bot: The color of the ball is blue.
<cogita-bot> Hello linas, you said: " The color of the ball is blue."
<cogita-bot> You made a Declarative SpeechAct
<linas> cogita-bot: What is the color of the ball?
<cogita-bot> Hello linas, you said: " What is the color of the ball?"
<cogita-bot> You made an Interrogative SpeechAct
<cogita-bot> the color of the ball is blue .
<linas> cogita-bot: What is the color of the book?
<cogita-bot> Hello linas, you said: " What is the color of the book?"
<cogita-bot> You made an Interrogative SpeechAct
<cogita-bot> the color of the ball is blue .
<linas> wtf
```
It knew the color of the book, but decided I'm now talking about the ball ??? It answered the wrong question.

Here is a different example: the adjectives are being ignored:
```
<linas> cogita-bot: John threw a green ball.
<cogita-bot> Hello linas, you said: " John threw a green ball."
<cogita-bot> You made a Declarative SpeechAct
<linas> cogita-bot: Who threw a ball?
<cogita-bot> Hello linas, you said: " Who threw a ball?"
<cogita-bot> You made an Interrogative SpeechAct
<cogita-bot> John threw a green ball .
<linas> cogita-bot: Mike threw a red ball.
<cogita-bot> Hello linas, you said: " Mike threw a red ball."
<cogita-bot> You made a Declarative SpeechAct
<linas> cogita-bot: Who threw a green ball?
<cogita-bot> Hello linas, you said: " Who threw a green ball?"
<cogita-bot> You made an Interrogative SpeechAct
<cogita-bot> Mike threw a red ball .
<linas> whoops
```

**Comment by linas** [comment]
Argh. It gets worse
```
scheme@(guile-user)> (process-query "luser" "John Kennedy died.")
Info: send to parser:  John Kennedy died. 
Info: close socket to parser
Hello luser, you said: "John Kennedy died."
You made a Declarative SpeechAct
scheme@(guile-user)> (process-query "luser" "Who died?")
Info: send to parser:  Who died? 
Info: close socket to parser
Hello luser, you said: "Who died?"
You made an Interrogative SpeechAct
Fuzzy match is finished.
$2 = "John Kennedy died ."
```
 Very good. continuing ... 
```
 scheme@(guile-user)> (process-query "luser" "What is the color of the ball? ")
Info: send to parser:  What is the color of the ball?  
Info: close socket to parser
Hello luser, you said: "What is the color of the ball? "
You made an Interrogative SpeechAct
Fuzzy match is finished.
$3 = "the color of the ball is blue ."
scheme@(guile-user)> (process-query "luser" "What is the color of the book? ")
Info: send to parser:  What is the color of the book?  
Info: close socket to parser
Hello luser, you said: "What is the color of the book? "
You made an Interrogative SpeechAct
Fuzzy match is finished.
$4 = "the color of the ball is blue ."
```
Answering the wrong question, as above.  Now get this:

```
scheme@(guile-user)> (process-query "luser" "Who died?")
Info: send to parser:  Who died? 
Info: close socket to parser
Hello luser, you said: "Who died?"
You made an Interrogative SpeechAct
Fuzzy match is finished.
$5 = "Linas ate an apple ."
```

Huh??? How'd we get from that question to that answer? Its not even close!

**Comment by leungmanhin** [comment]
Hmm... roughly the fuzzy matcher works like this, given a pattern (which in this case is the SetLink in that ReferenceLink that r2l generates for the input question), it searches for potential solutions. It compares each of the potential solutions with the pattern to see how many common nodes they share, finds out the (1 / incoming set size) of each of the common nodes, and calculates the similarity by adding them up, finally returns the most similar one. If there are multiple equally-similar results, it returns the one with the min size diff with the input. So in short the result not only depends on how "similar" the question and the answers are, but also depends on what else is in the atomspace. That may be one of the reasons.

Also the question and answer may share some node that's not a word of a sentence, for example `(DefinedLinguisticPredicateNode "definite")`, `(DefinedLinguisticConceptNode "past")` etc, but they contribute to the similarity-score, this may also be another reason why it sometimes returns irrelevant results. Anyway I'll take a deeper look at it.

**Comment by linas** [comment]
Yes please. Maybe we should have two fuzzy matchers: a "generic" one, for alter use and modification, as a "base class", and a "linguistic" one, which gives higher weight to _subj, _obj _adj, ...etc  relations, lower weight to tense and count relations.  The point being that adjectives and other modifiers are important for drawing distinctions, more than tense and count.

**Comment by leungmanhin** [comment]
Sure, perhaps I should start implementing that linguistic one as well

**Comment by linas** [comment]
See also issue #1922 -- I'm thinking that future versions could use the fuzzy matcher to get three or 4 answers, and then stitch them together.  For example:  "Able threw a ball" "Baker threw a ball" "Who throw a ball?  -- the fuzzy matcher should return both able and baker as matches, which then need to be merged before being sent off to microplanning.    Not clear how to do this yet, but that is for discussion in #1922 and not here.

Here, I just want to ask "who died?" and not get "Linas ate an apple" as the answer!

**Comment by linas** [comment]
I still have not studied the fuzzy matcher very much, so I am not clear on how it works, so do not have any practical advice at this time ...

**Comment by leungmanhin** [comment]
I think the fuzzy matcher is originally meant to find and return similar hypergraphs, to merge two or more results maybe we should implement something else outside of the fuzzy matcher to do so?

**Comment by linas** [comment]
yes. merging is NOT part of the fuzzy matcher! Should not be part of it.

But I should be able to ask the fuzzy matcher for 2 or 3 or 5 top-ranked results, and get them all, so that they could be merged.  There should also be a quality ranking.  so, for example "Able threw a ball gets a ranking of 0.9 and "baker threw a ball" gets a ranking of 0.9 and "linas threw an apple" get a ranking of 0.6 and "linas ate an apple" gets a ranking of 0.3  -- then I would (in some other code) merge all results with a score greater than 0.8 ...

**Comment by leungmanhin** [comment]
Yes sure, will also do that
2015年12月3日 13:46 於 "Linas Vepštas" <notifications@github.com> 寫道：

**Comment by linas** [comment]
OK much better! Looks like https://github.com/opencog/atomspace/pull/516 fixed whatever the issue was.  Closing.  Thanks!
